src/core/proxy/src/listener.rs

```diff
@@ -3,181 +3,295 @@
+
+        session.do_handshake()
+    }
+
+    /// handle a single session event
+    fn session_event(&mut self, event: &Event) {
         let token = event.token();
 
-        // handle error events first
         if event.is_error() {
             LISTENER_EVENT_ERROR.increment();
-            self.handle_error(token);
-        }
-
-        // handle write events before read events to reduce write
-        // buffer growth if there is also a readable event
-        if event.is_writable() {
-            LISTENER_EVENT_WRITE.increment();
-            self.do_write(token);
+            self.close(token);
+            return;
         }
 
-        // read events are handled last
         if event.is_readable() {
             LISTENER_EVENT_READ.increment();
-            let _ = self.do_read(token);
-        }
-
-        if let Ok(session) = self.poll.get_mut_session(token) {
-            if session.session.do_handshake().is_ok() {
-                trace!("handshake complete for session: {:?}", session.session);
-                if let Ok(session) = self.poll.remove_session(token) {
-                    if self
-                        .connection_queues
-                        .try_send_any(session.session)
-                        .is_err()
-                    {
-                        error!("error sending session to worker");
-                        TCP_ACCEPT_EX.increment();
-                    }
-                } else {
-                    error!("error removing session from poller");
-                    TCP_ACCEPT_EX.increment();
-                }
-            } else {
-                trace!("handshake incomplete for session: {:?}", session.session);
+            if self.read(token).is_err() {
+                self.close(token);
+                return;
             }
         }
-    }
 
-    pub fn do_accept(&mut self) {
-        if let Ok(token) = self.poll.accept() {
-            match self
-                .poll
-                .get_mut_session(token)
-                .map(|v| v.session.is_handshaking())
-            {
-                Ok(false) => {
-                    if let Ok(session) = self.poll.remove_session(token) {
-                        if self
-                            .connection_queues
-                            .try_send_any(session.session)
-                            .is_err()
-                        {
-                            warn!("rejecting connection, client connection queue is too full");
+        match self.handshake(token) {
+            Ok(_) => {
+                // handshake is complete, send the session to a worker thread
+                let mut session = self.sessions.remove(token.0);
+                for attempt in 1..=QUEUE_RETRIES {
+                    if let Err(s) = self.session_queue.try_send_any(session) {
+                        if attempt == QUEUE_RETRIES {
+                            LISTENER_SESSION_DISCARD.increment();
```

Comment:
We don't need to explicitly close. Since the session is removed from the slab on L#214, failure to move it onto the queue means it goes out of scope. Close happens automagically during drop.

Reply:
Comment added